letsencrypt/error_handler.py

```diff
@@ -0,0 +1,98 @@
+"""Registers functions to be called if an exception or signal occurs."""
+import logging
+import os
+import signal
+import traceback
+
+
+logger = logging.getLogger(__name__)
+
+
+# _SIGNALS stores the signals that will be handled by the ErrorHandler. These
+# signals were chosen as their default handler terminates the process and could
+# potentially occur from inside Python. Signals such as SIGILL were not
+# included as they could be a sign of something devious and we should terminate
+# immediately.
+_SIGNALS = ([signal.SIGTERM] if os.name == "nt" else
+            [signal.SIGTERM, signal.SIGHUP, signal.SIGQUIT,
+             signal.SIGXCPU, signal.SIGXFSZ, signal.SIGPWR])
+
+
+class ErrorHandler(object):
+    """Registers functions to be called if an exception or signal occurs.
+
+    This class allows you to register functions that will be called when
+    an exception or signal is encountered. The class works best as a
+    context manager. For example:
+
+    with ErrorHandler(cleanup_func):
+        do_something()
+
+    If an exception is raised out of do_something, cleanup_func will be
+    called. The exception is not caught by the ErrorHandler. Similarly,
+    if a signal is encountered, cleanup_func is called followed by the
+    previously registered signal handler.
+
+    Every registered function is attempted to be run to completion
+    exactly once. If a registered function raises an exception, it is
+    logged and the next function is called. If a (different) handled
+    signal occurs while calling a registered function, it is attempted
+    to be called again by the next signal handler.
```

Comment:
If you mean the following series of events, I agree there is a problem:

1. We're in the body of `call_registered`.
2. `self.funcs[-1]()` raises an exception.
3. We're inside the `except` block.
4. A signal is sent to the process.

In this case, we attempt to call `self.funcs[-1]` again because it was never popped from the list.

One issue with `pop` before calling is if `self.funcs[-1]` doesn't raise an exception but a signal interrupts it, `self.funcs[-1]` will mostly likely never complete. This is because `ErrorHandler._signal_handler` calls the previous signal handler before returning. Assuming this signal handler terminates the program, as is the case with the default signal handler of all signals in `_SIGNALS`, the handler of the signal that interrupted `self.funcs[-1]()` never returned so the function can never complete.